lib/Installation/Partitioner/LibstorageNG/v4_3/ExpertPartitionerController.pm

```diff
@@ -157,6 +157,10 @@ sub clone_partition_table {
 
 sub cancel_changes {
     my ($self, $args) = @_;
+    # Wait for menu bar to be shown to ensure that Expert Partitioner Wizard main page is opened,
+    # because 'Cancel' button exists on almost all the pages of the Wizard, and that may cause
+    # failures on slow systems while 'Cancel' is pressed on the wrong page.
+    $self->get_expert_partitioner_page()->wait_for_menu_bar_existence();
```

Comment:
Why not to do this check inline, as we don't use this method anywhere else?

Reply:
By inline do you mean using LibuiClient method in the Controller layer? It will break our agreement to not use library methods in Controller.

Reply:
I think it should go inside the getter, so we could check it before any action that a particular page is loaded, but I'm not sure if it is too early to consider that, like some method might call the getter not being in this screen, could you consider it if it is feasible?

Reply:
It may be feasible in case we would have it in all our getters. But I see an issue with that, because the wizard sometimes does not have unique elements, which can guarantee that the required page is opened (e.g. unique header). This solution will require some research and refactoring.

Also, even if I'll add it only in one getter for now, it will require a lot of testing, as `get_expert_partitioner_page()` is used a lot of places, and potentially may not have menu bar.

Reply:
ncurses won't have this bar, for instance. What we still could do is to ensure that warning is closed, for example. Is there maybe other element we could use and ensure it's existence when calling getter? Also, we already have https://progress.opensuse.org/issues/81248 so we could merge this solution and then design and test better way of handling such cases. I would also check if we can put unique id for those buttons in yast code, or they use some generic class.

Reply:
yes, that is my point, that this page perhaps is mature enough to use some short of ensure loading mechanism, as we have created multiple sub-page with other getters to handle other parts of the screen. it cannot be applied in a general way until we have more specific/derived pages. But yeah, might be out of scope, just thinking loud :)

Reply:
I was checking and I could only find the bar and the tree, the wizard has now some unset name not useful.

Reply:
@rwx788 

> What we still could do is to ensure that warning is closed, for example.

I was thinking about that solution too. But with that, the issue still may happen, because if the system is slow and popup is closed, then "Cancel" button will be immediately pressed, while new page is not loaded yet. The solution with checking if the correct page is present before pressing the button is more stable. 

> Is there maybe other element we could use and ensure it's existence when calling getter?
I've chosen the menu bar just because it indicates that the main page of Expert Partitioner Wizard is opened, regardless of the state of the right side of the page. I can use any other similar element for sure, but why the other element will be better?:)

> ncurses won't have this bar, for instance. 
Regarding ncurses, I think our tests will require another controller methods anyways, as the structure of the pages and elements are different sometimes if I correctly remember. Correct me if I'm wrong and our current methods could be used with ncurses without any changes. I guess, caring about this cross-environment ability is too far from our current implementation.

Reply:
We have plenty of unique controls there, like tree, menubar now, etc. So it's more about either checking that all of those are available or checking that previous dialog is closed and doesn't exist before pressing `cancel` button.

Reply:
Sorry, I would better discuss that in RocketChat, as I do not uderstand what is needed from me?:) I've described in the comment above why I did not use the solution with waiting for popup to be closed.

Reply:
To be clear, I've confused menubar with bargraph, which exists in qt only. So there is no issue with selected widget, but I still think that we can solve this problem not for a single case, but in general and ensure that correct dialog is displayed.

Reply:
I added the check to getter. Will test the solution and then will remove "WIP" when everything is ok.